### ISSUE TITLE
[Multi-device] SHLO->TTIR conversion fails when one of the input arguments to SHLO graph is a replicated scalar

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-mlir#3513]
### Describe the bug

This happens in context of manual data-parallel multichip MNIST training where we need learning rate replicated across all devices for local updates. For simplicity sake, below is SHLO graph for small repro code, which fails with error ```loc("data"): error: failed to legalize unresolved materialization from ('tensor<200x10xf32, #tt.mesh_sharding<"mesh", [ 2(1),  1]>>') to ('tensor<200x10xf32>') that remained live after conversion .```

```
module @jit_sharded_operation attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32} {
  sdy.mesh @mesh = <["dp"=2]>
  func.func public @main(%arg0: tensor<10x5xf32> {sdy.sharding = #sdy.sharding<@mesh, [{}, {}]>}, %arg1: tensor<200x10xf32> {sdy.sharding = #sdy.sharding<@mesh, [{"dp"}, {}]>}, %arg2: tensor<f32> {sdy.sharding = #sdy.sharding<@mesh, []>}) -> (tensor<f32> {jax.result_info = "", sdy.sharding = #sdy.sharding<@mesh, []>}) {
    %0 = sdy.manual_computation(%arg0, %arg1, %arg2) in_shardings=[<@mesh, [{}, {}]>, <@mesh, [{"dp"}, {}]>, <@mesh, []>] out_shardings=[<@mesh, []>] manual_axes={"dp"} (%arg3: tensor<10x5xf32>, %arg4: tensor<100x10xf32>, %arg5: tensor<f32>) {
      %cst = stablehlo.constant dense<1.000000e+03> : tensor<f32>
      %cst_0 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
      %1 = stablehlo.dot_general %arg4, %arg3, contracting_dims = [1] x [0] : (tensor<100x10xf32>, tensor<10x5xf32>) -> tensor<100x5xf32>
      %2 = stablehlo.broadcast_in_dim %arg5, dims = [] : (tensor<f32>) -> tensor<100x5xf32>
      %3 = stablehlo.multiply %1, %2 : tensor<100x5xf32>
      %4 = stablehlo.reshape %3 : (tensor<100x5xf32>) -> tensor<1x100x5xf32>
      %5 = "stablehlo.all_gather"(%4) <{all_gather_dim = 0 : i64, channel_handle = #stablehlo.channel_handle<handle = 1, type = 0>, replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>, use_global_device_ids}> : (tensor<1x100x5xf32>) -> tensor<2x100x5xf32>
      %6 = stablehlo.reshape %5 : (tensor<2x100x5xf32>) -> tensor<200x5xf32>
      %7 = stablehlo.reduce(%6 init: %cst_0) applies stablehlo.add across dimensions = [0, 1] : (tensor<200x5xf32>, tensor<f32>) -> tensor<f32>
      %8 = stablehlo.divide %7, %cst : tensor<f32>
      sdy.return %8 : tensor<f32>
    } : (tensor<10x5xf32>, tensor<200x10xf32>, tensor<f32>) -> tensor<f32>
    return %0 : tensor<f32>
  }
}
```
SHLO graph without mentioned scalar, which executes and gives good results, is below.

```
module @jit_sharded_operation attributes {mhlo.num_partitions = 2 : i32, mhlo.num_replicas = 1 : i32} {
  sdy.mesh @mesh = <["dp"=2]>
  func.func public @main(%arg0: tensor<10x5xf32> {sdy.sharding = #sdy.sharding<@mesh, [{}, {}]>}, %arg1: tensor<200x10xf32> {sdy.sharding = #sdy.sharding<@mesh, [{"dp"}, {}]>}) -> (tensor<f32> {jax.result_info = "", sdy.sharding = #sdy.sharding<@mesh, []>}) {
    %0 = sdy.manual_computation(%arg0, %arg1) in_shardings=[<@mesh, [{}, {}]>, <@mesh, [{"dp"}, {}]>] out_shardings=[<@mesh, []>] manual_axes={"dp"} (%arg2: tensor<10x5xf32>, %arg3: tensor<100x10xf32>) {
      %cst = stablehlo.constant dense<1.000000e+03> : tensor<f32>
      %cst_0 = stablehlo.constant dense<0.000000e+00> : tensor<f32>
      %1 = stablehlo.dot_general %arg3, %arg2, contracting_dims = [1] x [0] : (tensor<100x10xf32>, tensor<10x5xf32>) -> tensor<100x5xf32>
      %2 = stablehlo.reshape %1 : (tensor<100x5xf32>) -> tensor<1x100x5xf32>
      %3 = "stablehlo.all_gather"(%2) <{all_gather_dim = 0 : i64, channel_handle = #stablehlo.channel_handle<handle = 1, type = 0>, replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>, use_global_device_ids}> : (tensor<1x100x5xf32>) -> tensor<2x100x5xf32>
      %4 = stablehlo.reshape %3 : (tensor<2x100x5xf32>) -> tensor<200x5xf32>
      %5 = stablehlo.reduce(%4 init: %cst_0) applies stablehlo.add across dimensions = [0, 1] : (tensor<200x5xf32>, tensor<f32>) -> tensor<f32>
      %6 = stablehlo.divide %5, %cst : tensor<f32>
      sdy.return %6 : tensor<f32>
    } : (tensor<10x5xf32>, tensor<200x10xf32>) -> tensor<f32>
    return %0 : tensor<f32>
  }
}
```

**Comment by umalesTT** [comment]
FYI @wooseokTT

**Comment by wooseokTT** [comment]
@umalesTT  I did initial evaluation and can reproduce the problem. So far, I found that if one of the function arg is scalar, it leads to failure in creating a new op (i.e. mesh_shard in our case) mainly because of UNKNOWN SSA VALUE (arg1) issue. 

%1 = "ttir.mesh_shard"(<<UNKNOWN SSA VALUE>>, %0) <{shard_dims = array<i64: -1, 0>, shard_direction = #tt.shard_direction<full_to_shard>, shard_shape = array<i64: 2, 1>, shard_type = #tt.shard_type<identity>}> : (tensor<200x10xf32>, tensor<100x10xf32, #tt.mesh_sharding<"mesh">>) -> tensor<100x10xf32, #tt.mesh_sharding<"mesh">>

If we use your second graph, we can see no issue in identifying arg1.

%1 = "ttir.mesh_shard"(%arg1, %0) <{shard_dims = array<i64: -1, 0>, shard_direction = #tt.shard_direction<full_to_shard>, shard_shape = array<i64: 2, 1>, shard_type = #tt.shard_type<identity>}> : (tensor<200x10xf32, #tt.mesh_sharding<"mesh", [ 2(1),  1]>>, tensor<100x10xf32, #tt.mesh_sharding<"mesh">>) -> tensor<100x10xf32, #tt.mesh_sharding<"mesh">>

This also happens if I add dummy scalar arg to your second example graph. In this case, we are not touching newly added dummy arg. I also re-ordered arg1 and arg2 and it leads to the same issue.

Based on these observations, my current speculation is that it is not an issue of our code, but somewhere else such as inside of llvm or something. I will create a simple test case to see if we can reproduce this issue.

**Comment by wooseokTT** [comment]
https://github.com/tenstorrent/tt-mlir/pull/3520

**Comment by umalesTT** [comment]
I see, thanks!
I'll close the issue once the PR is merged.

**Comment by tapspatel** [comment]
pr was merged: https://github.com/tenstorrent/tt-mlir/pull/3520
